numpy/core/src/multiarray/ctors.c

```diff
@@ -1926,14 +1926,35 @@ PyArray_FromArray(PyArrayObject *arr, PyArray_Descr *newtype, int flags)
     /* Raise an error if the casting rule isn't followed */
     if (!PyArray_CanCastArrayTo(arr, newtype, casting)) {
```

Comment:
After some more debugging with `gdb` apparently the segmentation fault is occurring at this line.